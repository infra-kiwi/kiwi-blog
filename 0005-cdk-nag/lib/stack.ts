/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { aws_kms, aws_sns, CfnOutput } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class KiwiStack extends cdk.Stack {
  readonly topic: aws_sns.Topic;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    // Use the default AWS-provided SNS KMS key
    const awsSNSKey = aws_kms.Alias.fromAliasName(
      this,
      'AWSManagedSNSKey',
      'alias/aws/sns'
    );

    this.topic = new aws_sns.Topic(this, 'MyTopic', {
      masterKey: awsSNSKey
    });
    new CfnOutput(this, 'MyTopicArn', {
      value: this.topic.topicArn
    });
  }
}
