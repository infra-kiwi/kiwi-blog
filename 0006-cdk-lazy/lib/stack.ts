import * as cdk from 'aws-cdk-lib';
import { aws_sns, Stack } from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { MultiLineParameter } from './multi-line-parameter';

export class KiwiStack extends cdk.Stack {
  readonly topic: aws_sns.Topic;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    const myParameter = new MultiLineParameter(this, 'MyParameter', {
      parameterName: Stack.of(this).stackName + '-my-parameter'
    });

    // Populate the lines of our multi-line parameter
    myParameter.addLine("It's not DNS");
    myParameter.addLine("There's no way it's DNS");
    myParameter.addLine('It was DNS');

    // Add a dynamic line that depends on another CDK resource's value
    const aTopicAboutDNS = new aws_sns.Topic(this, 'ATopicAboutDNS');
    myParameter.addLine(
      `Here is a topic about DNS: ${aTopicAboutDNS.topicArn}`
    );
    this.topic = aTopicAboutDNS;
  }
}
