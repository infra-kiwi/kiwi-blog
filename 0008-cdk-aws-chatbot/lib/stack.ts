/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { Construct } from 'constructs';
import { ChatBotSlackChannel } from './constructs/chatbot-slack-channel';
import { PostToSlack } from './constructs/post-to-slack';
import { Config } from './util/config';

export type ChatBotStackProps = cdk.StackProps & {
  config: Config;
};

export class ChatBotStack extends cdk.Stack {
  constructor(scope: Construct, id: string, props: ChatBotStackProps) {
    super(scope, id, props);

    const chatBotSlackChannels: Record<string, ChatBotSlackChannel> = {};

    // Create all the possible ChatBot channel configurations. To do so,
    // extract all available Slack workspace/channel pairs, deduplicate them,
    // and create a ChatBotSlackChannel for each
    const slackChannelEntries = Array.from(
      new Set(
        Object.entries(props.config.notificationChannels).map(([, v]) =>
          JSON.stringify([v.slackWorkspaceId, v.slackChannelId])
        )
      )
    ).map((e) => JSON.parse(e)) as [string, string][];

    for (const entry of slackChannelEntries) {
      const [slackWorkspaceId, slackChannelId] = entry;

      // Configure the channel
      const channel = new ChatBotSlackChannel(
        this,
        `ChatBotSlackChannel-${slackWorkspaceId}-${slackChannelId}`,
        {
          slackWorkspaceId,
          slackChannelId
        }
      );
      chatBotSlackChannels[JSON.stringify(entry)] = channel;
    }

    // Create all the PostToSlack instances, which will accept our custom payloads
    // and translate them into ChatBot-formatted ones
    for (const notificationChannelName in props.config.notificationChannels) {
      const config = props.config.notificationChannels[notificationChannelName];

      // Use the previously-generated Slack channel configuration
      const chatBotSlackChannel =
        chatBotSlackChannels[
          JSON.stringify([config.slackWorkspaceId, config.slackChannelId])
        ];

      new PostToSlack(this, `SlackLambda-${notificationChannelName}`, {
        chatbotSNSTopicArn: chatBotSlackChannel.topicArn,
        notificationChannelName
      });
    }
  }
}
