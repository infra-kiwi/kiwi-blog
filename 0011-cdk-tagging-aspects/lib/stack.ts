/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import * as cdk from 'aws-cdk-lib';
import { aws_sns, CfnOutput } from 'aws-cdk-lib';
import { Construct } from 'constructs';

export class KiwiStack extends cdk.Stack {
  readonly topic: aws_sns.Topic;

  constructor(scope: Construct, id: string, props?: cdk.StackProps) {
    super(scope, id, props);

    this.topic = new aws_sns.Topic(this, 'MyTopic');

    // Define an output for debugging purposes
    new CfnOutput(this, 'MyTopicArnOutput', {
      value: this.topic.topicArn
    });
  }
}
