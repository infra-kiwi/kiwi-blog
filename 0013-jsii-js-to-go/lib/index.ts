/*
 * Copyright (c) 2025. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

export class Constants {
  static readonly name = 'Mario';
}

export interface Home {
  readonly address: string;
}

export class Addressor {
  getFullGreeting(name: string, home: Home) {
    return `Hello ${Constants.name}! You live at ${home.address}.`;
  }
}
