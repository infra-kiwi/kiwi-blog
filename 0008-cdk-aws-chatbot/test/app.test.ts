/*
 * Copyright (c) 2024. Alberto Marchetti - https://www.linkedin.com/in/albertomarchetti/
 */

import { ChatBotStack } from '../lib/stack';
import { App } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';

test('ChatBotStack', async () => {
  const app = new App();

  // Init stack
  const stack = new ChatBotStack(app, 'Test', {
    config: {
      notificationChannels: {
        'my-channel': {
          slackWorkspaceId: 'T1234',
          slackChannelId: 'C1234'
        },
        'my-channel-2': {
          slackWorkspaceId: 'T1234',
          slackChannelId: 'C5678'
        },
        'my-channel-3': {
          // Repeated workspace/channel pair
          slackWorkspaceId: 'T1234',
          slackChannelId: 'C1234'
        }
      }
    }
  });

  // Prepare for assertions
  const template = Template.fromStack(stack);
  expect(template).toMatchSnapshot();

  template.resourceCountIs(
    'AWS::SNS::Topic',
    // 2 ChatBot + 3 ingestion
    5
  );
  // Own lambda + ChatBot log retention lambda
  template.resourceCountIs(
    'AWS::Lambda::Function',
    // 1 ChatBot log retention + 3 ingestion
    4
  );
});
