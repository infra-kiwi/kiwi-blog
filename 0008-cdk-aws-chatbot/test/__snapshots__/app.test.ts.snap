// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ChatBotStack 1`] = `
{
  "Mappings": {
    "LatestNodeRuntimeMap": {
      "af-south-1": {
        "value": "nodejs20.x",
      },
      "ap-east-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-1": {
        "value": "nodejs20.x",
      },
      "ap-northeast-2": {
        "value": "nodejs20.x",
      },
      "ap-northeast-3": {
        "value": "nodejs20.x",
      },
      "ap-south-1": {
        "value": "nodejs20.x",
      },
      "ap-south-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-1": {
        "value": "nodejs20.x",
      },
      "ap-southeast-2": {
        "value": "nodejs20.x",
      },
      "ap-southeast-3": {
        "value": "nodejs20.x",
      },
      "ap-southeast-4": {
        "value": "nodejs20.x",
      },
      "ap-southeast-5": {
        "value": "nodejs20.x",
      },
      "ap-southeast-7": {
        "value": "nodejs20.x",
      },
      "ca-central-1": {
        "value": "nodejs20.x",
      },
      "ca-west-1": {
        "value": "nodejs20.x",
      },
      "cn-north-1": {
        "value": "nodejs18.x",
      },
      "cn-northwest-1": {
        "value": "nodejs18.x",
      },
      "eu-central-1": {
        "value": "nodejs20.x",
      },
      "eu-central-2": {
        "value": "nodejs20.x",
      },
      "eu-isoe-west-1": {
        "value": "nodejs18.x",
      },
      "eu-north-1": {
        "value": "nodejs20.x",
      },
      "eu-south-1": {
        "value": "nodejs20.x",
      },
      "eu-south-2": {
        "value": "nodejs20.x",
      },
      "eu-west-1": {
        "value": "nodejs20.x",
      },
      "eu-west-2": {
        "value": "nodejs20.x",
      },
      "eu-west-3": {
        "value": "nodejs20.x",
      },
      "il-central-1": {
        "value": "nodejs20.x",
      },
      "me-central-1": {
        "value": "nodejs20.x",
      },
      "me-south-1": {
        "value": "nodejs20.x",
      },
      "mx-central-1": {
        "value": "nodejs20.x",
      },
      "sa-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-1": {
        "value": "nodejs20.x",
      },
      "us-east-2": {
        "value": "nodejs20.x",
      },
      "us-gov-east-1": {
        "value": "nodejs18.x",
      },
      "us-gov-west-1": {
        "value": "nodejs18.x",
      },
      "us-iso-east-1": {
        "value": "nodejs18.x",
      },
      "us-iso-west-1": {
        "value": "nodejs18.x",
      },
      "us-isob-east-1": {
        "value": "nodejs18.x",
      },
      "us-west-1": {
        "value": "nodejs20.x",
      },
      "us-west-2": {
        "value": "nodejs20.x",
      },
    },
  },
  "Outputs": {
    "ChatBotSlackChannelChatBotSlackChannelT1234C1234ConfigurationUrl0ADB8ABA": {
      "Value": "https://console.aws.amazon.com/chatbot/home?#/chat-clients/slack/workspaces/T1234/configurations/slack-T1234-C1234",
    },
    "ChatBotSlackChannelChatBotSlackChannelT1234C1234TopicArn6B64EFF3": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "arn:",
            {
              "Ref": "AWS::Partition",
            },
            ":sns:",
            {
              "Ref": "AWS::Region",
            },
            ":",
            {
              "Ref": "AWS::AccountId",
            },
            ":Test-chatbot-raw-T1234-C1234",
          ],
        ],
      },
    },
    "ChatBotSlackChannelChatBotSlackChannelT1234C5678ConfigurationUrlFDF9648B": {
      "Value": "https://console.aws.amazon.com/chatbot/home?#/chat-clients/slack/workspaces/T1234/configurations/slack-T1234-C5678",
    },
    "ChatBotSlackChannelChatBotSlackChannelT1234C5678TopicArn41353A77": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "arn:",
            {
              "Ref": "AWS::Partition",
            },
            ":sns:",
            {
              "Ref": "AWS::Region",
            },
            ":",
            {
              "Ref": "AWS::AccountId",
            },
            ":Test-chatbot-raw-T1234-C5678",
          ],
        ],
      },
    },
    "PostToSlackSlackLambdamychannel2TopicArn6FD825F5": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "arn:",
            {
              "Ref": "AWS::Partition",
            },
            ":sns:",
            {
              "Ref": "AWS::Region",
            },
            ":",
            {
              "Ref": "AWS::AccountId",
            },
            ":chatbot-my-channel-2",
          ],
        ],
      },
    },
    "PostToSlackSlackLambdamychannel3TopicArn44107F85": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "arn:",
            {
              "Ref": "AWS::Partition",
            },
            ":sns:",
            {
              "Ref": "AWS::Region",
            },
            ":",
            {
              "Ref": "AWS::AccountId",
            },
            ":chatbot-my-channel-3",
          ],
        ],
      },
    },
    "PostToSlackSlackLambdamychannelTopicArn1ABD0783": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "arn:",
            {
              "Ref": "AWS::Partition",
            },
            ":sns:",
            {
              "Ref": "AWS::Region",
            },
            ":",
            {
              "Ref": "AWS::AccountId",
            },
            ":chatbot-my-channel",
          ],
        ],
      },
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ChatBotSlackChannelChatBotSlackChannelT1234C1234SNSTopicB65800E8": {
      "Properties": {
        "TopicName": "Test-chatbot-raw-T1234-C1234",
      },
      "Type": "AWS::SNS::Topic",
    },
    "ChatBotSlackChannelChatBotSlackChannelT1234C1234SlackChannel371F35AC": {
      "Properties": {
        "ConfigurationName": "slack-T1234-C1234",
        "IamRoleArn": {
          "Fn::GetAtt": [
            "ChatBotSlackChannelChatBotSlackChannelT1234C1234SlackChannelConfigurationRole6D8470B6",
            "Arn",
          ],
        },
        "LoggingLevel": "INFO",
        "SlackChannelId": "C1234",
        "SlackWorkspaceId": "T1234",
        "SnsTopicArns": [
          {
            "Ref": "ChatBotSlackChannelChatBotSlackChannelT1234C1234SNSTopicB65800E8",
          },
        ],
      },
      "Type": "AWS::Chatbot::SlackChannelConfiguration",
    },
    "ChatBotSlackChannelChatBotSlackChannelT1234C1234SlackChannelConfigurationRole6D8470B6": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "chatbot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "ChatBotSlackChannelChatBotSlackChannelT1234C1234SlackChannelLogRetentionCC3F095A": {
      "Properties": {
        "LogGroupName": "/aws/chatbot/slack-T1234-C1234",
        "LogGroupRegion": "us-east-1",
        "RetentionInDays": 30,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "ChatBotSlackChannelChatBotSlackChannelT1234C5678SNSTopic255E9BF3": {
      "Properties": {
        "TopicName": "Test-chatbot-raw-T1234-C5678",
      },
      "Type": "AWS::SNS::Topic",
    },
    "ChatBotSlackChannelChatBotSlackChannelT1234C5678SlackChannel2699C235": {
      "Properties": {
        "ConfigurationName": "slack-T1234-C5678",
        "IamRoleArn": {
          "Fn::GetAtt": [
            "ChatBotSlackChannelChatBotSlackChannelT1234C5678SlackChannelConfigurationRoleDC1CA40E",
            "Arn",
          ],
        },
        "LoggingLevel": "INFO",
        "SlackChannelId": "C5678",
        "SlackWorkspaceId": "T1234",
        "SnsTopicArns": [
          {
            "Ref": "ChatBotSlackChannelChatBotSlackChannelT1234C5678SNSTopic255E9BF3",
          },
        ],
      },
      "Type": "AWS::Chatbot::SlackChannelConfiguration",
    },
    "ChatBotSlackChannelChatBotSlackChannelT1234C5678SlackChannelConfigurationRoleDC1CA40E": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "chatbot.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
      },
      "Type": "AWS::IAM::Role",
    },
    "ChatBotSlackChannelChatBotSlackChannelT1234C5678SlackChannelLogRetention8C4204AF": {
      "Properties": {
        "LogGroupName": "/aws/chatbot/slack-T1234-C5678",
        "LogGroupRegion": "us-east-1",
        "RetentionInDays": 30,
        "ServiceToken": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A",
            "Arn",
          ],
        },
      },
      "Type": "Custom::LogRetention",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aFD4BFC8A": {
      "DependsOn": [
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "2819175352ad1ce0dae768e83fc328fb70fb5f10b4a8ff0ccbcb791f02b0716d.zip",
        },
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
            "Arn",
          ],
        },
        "Runtime": {
          "Fn::FindInMap": [
            "LatestNodeRuntimeMap",
            {
              "Ref": "AWS::Region",
            },
            "value",
          ],
        },
        "Timeout": 900,
      },
      "Type": "AWS::Lambda::Function",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": [
                "logs:PutRetentionPolicy",
                "logs:DeleteRetentionPolicy",
              ],
              "Effect": "Allow",
              "Resource": "*",
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRoleDefaultPolicyADDA7DEB",
        "Roles": [
          {
            "Ref": "LogRetentionaae0aa3c5b4d4f87b02d85b201efdd8aServiceRole9741ECFB",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PostToSlackSlackLambdamychannel2SNSTopicD770D013": {
      "Properties": {
        "TopicName": "chatbot-my-channel-2",
      },
      "Type": "AWS::SNS::Topic",
    },
    "PostToSlackSlackLambdamychannel2SlackLambda51F3385A": {
      "DependsOn": [
        "PostToSlackSlackLambdamychannel2SlackLambdaServiceRoleDefaultPolicyEF945B0E",
        "PostToSlackSlackLambdamychannel2SlackLambdaServiceRole88899726",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "69648fa6d2afaa97402c4df6b21dbf01e48dc1b7e1f5c67d08b579067cf88f62.zip",
        },
        "Environment": {
          "Variables": {
            "SNS_TOPIC_ARN": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":sns:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":Test-chatbot-raw-T1234-C5678",
                ],
              ],
            },
          },
        },
        "FunctionName": "Test-slack-chatbot-my-channel-2",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "PostToSlackSlackLambdamychannel2SlackLambdaServiceRole88899726",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "PostToSlackSlackLambdamychannel2SlackLambdaAllowInvokeTestPostToSlackSlackLambdamychannel2SNSTopicD1F8D2C808E15E4F": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "PostToSlackSlackLambdamychannel2SlackLambda51F3385A",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "PostToSlackSlackLambdamychannel2SNSTopicD770D013",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "PostToSlackSlackLambdamychannel2SlackLambdaSNSTopic6DA2D742": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "PostToSlackSlackLambdamychannel2SlackLambda51F3385A",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "PostToSlackSlackLambdamychannel2SNSTopicD770D013",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "PostToSlackSlackLambdamychannel2SlackLambdaServiceRole88899726": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PostToSlackSlackLambdamychannel2SlackLambdaServiceRoleDefaultPolicyEF945B0E": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":sns:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":Test-chatbot-raw-T1234-C5678",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PostToSlackSlackLambdamychannel2SlackLambdaServiceRoleDefaultPolicyEF945B0E",
        "Roles": [
          {
            "Ref": "PostToSlackSlackLambdamychannel2SlackLambdaServiceRole88899726",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PostToSlackSlackLambdamychannel3SNSTopic5D95AABB": {
      "Properties": {
        "TopicName": "chatbot-my-channel-3",
      },
      "Type": "AWS::SNS::Topic",
    },
    "PostToSlackSlackLambdamychannel3SlackLambdaAllowInvokeTestPostToSlackSlackLambdamychannel3SNSTopic53A6C2E1BA94893B": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "PostToSlackSlackLambdamychannel3SlackLambdaFE36F36B",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "PostToSlackSlackLambdamychannel3SNSTopic5D95AABB",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "PostToSlackSlackLambdamychannel3SlackLambdaFE36F36B": {
      "DependsOn": [
        "PostToSlackSlackLambdamychannel3SlackLambdaServiceRoleDefaultPolicyF91082F7",
        "PostToSlackSlackLambdamychannel3SlackLambdaServiceRoleDF7C4C06",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "69648fa6d2afaa97402c4df6b21dbf01e48dc1b7e1f5c67d08b579067cf88f62.zip",
        },
        "Environment": {
          "Variables": {
            "SNS_TOPIC_ARN": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":sns:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":Test-chatbot-raw-T1234-C1234",
                ],
              ],
            },
          },
        },
        "FunctionName": "Test-slack-chatbot-my-channel-3",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "PostToSlackSlackLambdamychannel3SlackLambdaServiceRoleDF7C4C06",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "PostToSlackSlackLambdamychannel3SlackLambdaSNSTopic4567FD95": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "PostToSlackSlackLambdamychannel3SlackLambdaFE36F36B",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "PostToSlackSlackLambdamychannel3SNSTopic5D95AABB",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "PostToSlackSlackLambdamychannel3SlackLambdaServiceRoleDF7C4C06": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PostToSlackSlackLambdamychannel3SlackLambdaServiceRoleDefaultPolicyF91082F7": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":sns:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":Test-chatbot-raw-T1234-C1234",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PostToSlackSlackLambdamychannel3SlackLambdaServiceRoleDefaultPolicyF91082F7",
        "Roles": [
          {
            "Ref": "PostToSlackSlackLambdamychannel3SlackLambdaServiceRoleDF7C4C06",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
    "PostToSlackSlackLambdamychannelSNSTopic6F5F2942": {
      "Properties": {
        "TopicName": "chatbot-my-channel",
      },
      "Type": "AWS::SNS::Topic",
    },
    "PostToSlackSlackLambdamychannelSlackLambda4496A26D": {
      "DependsOn": [
        "PostToSlackSlackLambdamychannelSlackLambdaServiceRoleDefaultPolicy80E5755D",
        "PostToSlackSlackLambdamychannelSlackLambdaServiceRole75730855",
      ],
      "Properties": {
        "Code": {
          "S3Bucket": {
            "Fn::Sub": "cdk-hnb659fds-assets-\${AWS::AccountId}-\${AWS::Region}",
          },
          "S3Key": "69648fa6d2afaa97402c4df6b21dbf01e48dc1b7e1f5c67d08b579067cf88f62.zip",
        },
        "Environment": {
          "Variables": {
            "SNS_TOPIC_ARN": {
              "Fn::Join": [
                "",
                [
                  "arn:",
                  {
                    "Ref": "AWS::Partition",
                  },
                  ":sns:",
                  {
                    "Ref": "AWS::Region",
                  },
                  ":",
                  {
                    "Ref": "AWS::AccountId",
                  },
                  ":Test-chatbot-raw-T1234-C1234",
                ],
              ],
            },
          },
        },
        "FunctionName": "Test-slack-chatbot-my-channel",
        "Handler": "index.handler",
        "Role": {
          "Fn::GetAtt": [
            "PostToSlackSlackLambdamychannelSlackLambdaServiceRole75730855",
            "Arn",
          ],
        },
        "Runtime": "nodejs20.x",
      },
      "Type": "AWS::Lambda::Function",
    },
    "PostToSlackSlackLambdamychannelSlackLambdaAllowInvokeTestPostToSlackSlackLambdamychannelSNSTopic798371FB5206BF55": {
      "Properties": {
        "Action": "lambda:InvokeFunction",
        "FunctionName": {
          "Fn::GetAtt": [
            "PostToSlackSlackLambdamychannelSlackLambda4496A26D",
            "Arn",
          ],
        },
        "Principal": "sns.amazonaws.com",
        "SourceArn": {
          "Ref": "PostToSlackSlackLambdamychannelSNSTopic6F5F2942",
        },
      },
      "Type": "AWS::Lambda::Permission",
    },
    "PostToSlackSlackLambdamychannelSlackLambdaSNSTopicF0A53D2E": {
      "Properties": {
        "Endpoint": {
          "Fn::GetAtt": [
            "PostToSlackSlackLambdamychannelSlackLambda4496A26D",
            "Arn",
          ],
        },
        "Protocol": "lambda",
        "TopicArn": {
          "Ref": "PostToSlackSlackLambdamychannelSNSTopic6F5F2942",
        },
      },
      "Type": "AWS::SNS::Subscription",
    },
    "PostToSlackSlackLambdamychannelSlackLambdaServiceRole75730855": {
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "lambda.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AWSLambdaBasicExecutionRole",
              ],
            ],
          },
        ],
      },
      "Type": "AWS::IAM::Role",
    },
    "PostToSlackSlackLambdamychannelSlackLambdaServiceRoleDefaultPolicy80E5755D": {
      "Properties": {
        "PolicyDocument": {
          "Statement": [
            {
              "Action": "sns:Publish",
              "Effect": "Allow",
              "Resource": {
                "Fn::Join": [
                  "",
                  [
                    "arn:",
                    {
                      "Ref": "AWS::Partition",
                    },
                    ":sns:",
                    {
                      "Ref": "AWS::Region",
                    },
                    ":",
                    {
                      "Ref": "AWS::AccountId",
                    },
                    ":Test-chatbot-raw-T1234-C1234",
                  ],
                ],
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "PolicyName": "PostToSlackSlackLambdamychannelSlackLambdaServiceRoleDefaultPolicy80E5755D",
        "Roles": [
          {
            "Ref": "PostToSlackSlackLambdamychannelSlackLambdaServiceRole75730855",
          },
        ],
      },
      "Type": "AWS::IAM::Policy",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
